### PR TITLE
Fix scene deletion bug & prevent deletion of final scene

### PR DESCRIPTION
**Pull request by grant-davidson**
Currently, deleting the last scene has no effect because claymate immediately adds the scene back again based on the current drawing.

Users would expect that deleting the last scene results in an empty drawing.

### CODE REVIEW

**Comment by dai-shi** [comment]
Hm? The preview doesn't seem to change the behavior from production.

> Users would expect that deleting the last scene results in an empty drawing.

For me, the current behavior on production is intuitive.

**Comment by grant-davidson** [comment]
> Hm? The preview doesn't seem to change the behavior from production.

It does for me.  There should be 2 changes:

- The correct scene is deleted (production deletes the scene after the one that should be deleted)
- Deleting the last scene should clear the drawing

Which browser are you using?

> For me, the current behavior on production is intuitive.

How about I modify the PR so that on the final scene the delete button is disabled?  In the production version it seems wrong to have a delete button that does nothing but cause the window to flicker

**Comment by dai-shi** [comment]
>     * The correct scene is deleted (production deletes the scene after the one that should be deleted)

Oh, I see. Yeah, it deletes the one after the one clicked in production.
The preview fixes it.

Interestingly, in production if you click the last one for the first time, nothing happens, but clicking again will delete it.

>     * Deleting the last scene should clear the drawing

I'm not sure what "clear" means. The behavior in preview is deleting the last one deletes it and shows the previous one, which I think is correct/intuitive.

> Which browser are you using?

Chrome/MacOS

> How about I modify the PR so that on the final scene the delete button is disabled?

I think we would like to delete the final scene.

**Comment by dai-shi** [comment]
we'd want to keep the selection when deleting a scene. The screen recording shows you are on slide 1 and after deleting slide 3, it moves to slide 2.


https://user-images.githubusercontent.com/490574/117370817-eb279500-af01-11eb-83fe-4376c49c12c2.mov

File: src/Claymate.tsx [hunk 1]
```diff
@@ -119,6 +118,7 @@ const Claymate: React.FC<Props> = ({
               type="button"
               className="Claymate-delete"
               aria-label="Delete"
+              disabled={scenes.length <= 1}
```
Comment (dai-shi) [inline src/Claymate.tsx:121]: 👍

File: src/Claymate.tsx [hunk 2]
```diff
@@ -46,16 +46,15 @@ const Claymate: React.FC<Props> = ({
...
       const remainingScenes = scenes.length - 1;
-      let newIndex;
       if (remainingScenes > 0) {
-        newIndex = index < remainingScenes ? index : remainingScenes - 1;
+        const newIndex = index < remainingScenes ? index : remainingScenes - 1;
+        const drawingSourceIndex =
+          index < remainingScenes ? index + 1 : index - 1;
+        updateScenes((prev: Scene[]) => prev.filter((item) => item.id !== id), {
+          index: newIndex,
+          drawing: scenes[drawingSourceIndex].drawing,
```
Comment (dai-shi) [inline src/Claymate.tsx:55]: I'm not sure if I understand why we need two index numbers.

Comment (grant-davidson) [reply src/Claymate.tsx]: The source of the new scene after deletion is not necessarily based on the index that will be current after the deletion

Comment (dai-shi) [reply src/Claymate.tsx]: > The source of the new scene after deletion is not necessarily based on the index that will be current after the deletion

Oh, I see, sure. That was my misunderstanding. Thanks for explanation.

**Comment by dai-shi** [comment]
Thanks for your work, but the behavior is strange. This is unexpected, right?

https://user-images.githubusercontent.com/490574/117527712-fdd8c180-b008-11eb-83c2-dbf26aa2573e.mov

**Comment by grant-davidson** [comment]
My mistake!  I need to find a way to test the behaviour.

File: src/Claymate.tsx [hunk 3]
```diff
@@ -46,16 +47,30 @@ const Claymate: React.FC<Props> = ({
...
+              if (currentIndex === remainingScenes) {
+                sourceIndex = currentIndex - 1;
+              } else if (currentIndex === 0) {
+                sourceIndex = 1;
+              }
+            }
+            newCurrent = {
+              index: currentIndex > 0 ? currentIndex - 1 : currentIndex,
+              drawing: scenes[sourceIndex].drawing,
+            };
```
Comment (dai-shi) [inline src/Claymate.tsx:66]: This part looks good to me.

**Review by dai-shi — APPROVED** [review]
I didn't check everything in detail, but looks good. We might find some ideas to improve it later.

Merging this. Thanks very much for your work.

**Pull request merged**